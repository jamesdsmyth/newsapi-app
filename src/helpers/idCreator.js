// idCreator converts the article title in a url friendly link
export function idCreator(str) {
  return str.replace(/[\W_]+/g,"-").toLowerCase();
}