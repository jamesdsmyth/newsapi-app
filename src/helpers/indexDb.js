import { openDB } from 'idb';

// save data locally by opening an index db and writing the new articles to the db
export async function saveCategoryDataLocally(article, category) {
  let db;
  try {
    db = await openDB('newsApp', 1, {
      upgrade(db) {
        db.createObjectStore('articles', {
          keyPath: 'title',
          autoIncrement: true,
        });
      }
    });
  } catch(error) {
    console.log(error);
  }

  // Uncaught (in promise) DOMException
  // not sure why but this keeps dropping into the catch error even though it writes to the IndexDB store.
  try {
    await db.add('articles', {
      title: category,
      body: article,
    })
  } catch(error) {
    console.log(error);
  }
}

// get the local data from the indexDB if it exists
export function* getLocalEventData(category) {
  if (!('indexedDB' in window)) {
    return;
  }
  
  const db = yield openDB('newsApp', 1);

  // if the IndexDB does not have an createObjectStore then we will return.
  if(!db || db.objectStoreNames.length === 0) {
    return;
  }

  const transaction = yield db.transaction('articles', 'readonly');
  const store = yield transaction.objectStore('articles');
  const value = yield store.getAll();
  const val = value.filter(item => item.title === category);

  return val[0];
}