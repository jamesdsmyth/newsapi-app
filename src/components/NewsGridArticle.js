import React from 'react';
import { Link } from 'react-router-dom';
import PropTypes from 'prop-types';

import { idCreator } from '../helpers/idCreator';

const NewsGridArticle = ({ article, category, index }) => {
  
  const articleId = idCreator(article.title);
  return (
    <article className={`article article--${index}`}>
      <div className="article__article-image-container">
        <img
          className="article__article-image"
          src={article.urlToImage} alt={article.title} 
        />
      </div>
      <div className="article__article-heading-container">
        <span className="article__article-source">
          {article.source.name}
        </span>
        <h2 className="article__article-heading">
          <Link 
            to={`${category}/${articleId}`}
            className="article__article-heading-link"
          >
            {article.title}
          </Link>
        </h2>
      </div>
    </article>
  )
}

NewsGridArticle.propTypes = {
  articles: PropTypes.array,
  category: PropTypes.string,
  index: PropTypes.number
}

export default NewsGridArticle;