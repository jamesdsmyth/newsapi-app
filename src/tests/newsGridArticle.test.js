import React from 'react';
import { shallow, configure } from 'enzyme';
import sinon from 'sinon';
import { BrowserRouter as Router } from 'react-router-dom';

import NewsGridArticle from '../components/NewsGridArticle';
import Adapter from 'enzyme-adapter-react-16';
import renderer from 'react-test-renderer';

import mockApiResponseSuccess from '../mockData/mock-api-success';
import { idCreator } from '../helpers/idCreator';

configure({ adapter: new Adapter() });

describe('<NewsGridArticle />', () => {
  const article = mockApiResponseSuccess.articles[0];
  const category = 'general'
  const index = 0;
  const articleId = idCreator(article.title);
  
  const wrapper = shallow(
    <NewsGridArticle 
      article={article} 
      category={category} 
      index={index} 
    />
  ); 
  
  it('should render the NewsGridArticle', () => {
    expect(wrapper.find('NewsGridArticle')).toBeDefined();
  });

  it('should have the correct wrapping class', () => {
    expect(wrapper.find('.article')).toBeDefined();
  });

  it('should have the correct wrapping index class', () => {
    expect(wrapper.find(`.article--${index}`)).toBeDefined();
  });

  it('should render an <img />', () => {
    expect(wrapper.find('img.article__article-image')).toBeDefined();
  });

  it('should render the <Img /> with the correct image', () => {
    expect(wrapper.find('img.article__article-image').props().src).toBe(article.urlToImage);
   });

  it('should render a <h2 />', () => {
    expect(wrapper.find('h2')).toBeDefined();
  });

  it('should render <span />', () => {
    expect(wrapper.find('.article__article-source')).toBeDefined();
  });

  it('should render <span /> with text of the passed source', () => {
    expect(wrapper.find('.article__article-source').text()).toEqual(article.source.name);
  });

  it('should render a <Link />', () => {
    expect(wrapper.find('Link')).toBeDefined();
  });
  
  it('should render the <Link /> with the correct href', () => {
    expect(wrapper.find('Link').props().to).toBe(`${category}/${articleId}`);
  });

  it('should render the <Link /> with the correct class', () => {
    expect(wrapper.find('Link').props().className).toBe('article__article-heading-link');
  });

  it('renders <NewsGridArticle /> and matches the snapshot', () => {
    const component = renderer.create(
      <Router>
        <NewsGridArticle 
          article={article} 
          category={category} 
          index={index} 
        />
      </Router>
    );
  
    expect(component.toJSON()).toMatchSnapshot();
  });

  it('should simulate the click event and expect an anonymous function', () => {
    const mockCallBack = sinon.spy();
    
    wrapper.find('Link').simulate('click');
    expect(mockCallBack).toEqual(expect.any(Function));
  });
});