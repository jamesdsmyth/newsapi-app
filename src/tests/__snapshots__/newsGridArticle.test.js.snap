// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`<NewsGridArticle /> renders <NewsGridArticle /> and matches the snapshot 1`] = `
<article
  className="article article--0"
>
  <div
    className="article__article-image-container"
  >
    <img
      alt="Maduro: 'Coup attempt' defeated as police clash with protesters - Al Jazeera English"
      className="article__article-image"
      src="https://www.aljazeera.com/mritems/Images/2019/5/1/f8cc2c622be846b5b049f09206901f1a_18.jpg"
    />
  </div>
  <div
    className="article__article-heading-container"
  >
    <span
      className="article__article-source"
    >
      Al Jazeera English
    </span>
    <h2
      className="article__article-heading"
    >
      <a
        className="article__article-heading-link"
        href="/general/maduro-coup-attempt-defeated-as-police-clash-with-protesters-al-jazeera-english"
        onClick={[Function]}
      >
        Maduro: 'Coup attempt' defeated as police clash with protesters - Al Jazeera English
      </a>
    </h2>
  </div>
</article>
`;
